build-std: always (#3874)
rust-lang/rfcs: text/3874-build-std-always.md

`.config/cargo.toml` in the project root and/or Cargo home directory), the value
of `build-std` is not influenced by the dependencies of the current crate.

When `build-std` is set to "always", then the standard library will be
unconditionally recompiled ([?][rationale-unconditional]) in the release profile
defined in its workspace as part of every clean build
([?][rationale-release-profile]). This is primarily useful for users of tier
three targets. As with other dependencies, the standard library's build will
respect the `RUSTFLAGS` environment variable.

> [!NOTE]
>
> Configuration of the pre-built standard library is split across bootstrap and
> the Cargo packages for the standard library. As much of this configuration as
> possible should be moved to the Cargo profile for these packages so that the
> artifacts produced by build-std match the pre-built standard library as much
> as is feasible.

Alongside `build-std`, a `build-std-crates` key will be introduced
([?][rationale-build-std-crate]), which can be used to specify which crates from
the standard library should be built. Only "core", "alloc" and "std" are valid
values for `build-std-crates`.

```toml
[build]

> build-std-crates = "std"

Reviewer: A related thought I had when seeing this field by itself is it could be interpreted to mean `build-std = "always"` on its own.  Having this example be more complete would help but it could still be confusing when layering different configs.

Unsure if its fine on its own, needs a different structure, a warning or note, or something else.

- [*Why use the pre-built standard library for procedural macros and build scripts in cross-compile mode?*][rationale-host-deps-cross]

*See the following sections for relevant unresolved questions:*

- [*What should the `build-std` configuration in `.cargo/config` be named?*][unresolved-config-name]
- [*What should the "always" and "never" values of `build-std` be named?*][unresolved-config-values]
- [*What should `build-std-crates` be named?*][unresolved-build-std-crate-name]
- [*Should the standard library inherit RUSTFLAGS?*][unresolved-inherit-rustflags]

*See the following sections for future possibilities:*

- [*Allow reusing sysroot artifacts if available*][future-reuse-sysroot]

## Standard library crate stability
[standard-library-crate-stability]: #standard-library-crate-stability

An optional `standard_library_support` field
([?][rationale-why-standard-library-support]) is added to the target
specification ([?][rationale-target-spec-purpose]), replacing the existing
`metadata.std` field. `standard_library_support` has two fields:

- `supported`, which can be set to either "core", "core and alloc", or

>   "core, alloc and std"

Author: Fixed in 3f1b89b

sources to this location and will be made a default component. If the sources
are not found, Cargo will emit an error and recommend the user download
`rust-src` if using rustup.

`rust-src` will contain the sources for the standard library crates as well as
its vendored dependencies ([?][rationale-vendoring]). As a consequence sources
of standard library dependencies will not need be fetched from crates.io.

> [!NOTE]
>
> Cargo will not perform any checks to ensure that the sources in `rust-src`
> have been modified ([?][rationale-src-modifications]). It will be documented
> that modifying these sources is not supported.

*See the following sections for rationale/alternatives:*

- [*Why not allow the source path for the standard library be customised?*][rationale-custom-src-path]
- [*Why vendor standard library dependencies?*][rationale-vendoring]
- [*Why not check if `rust-src` has been modified?*][rationale-src-modifications]

*See the following sections for relevant unresolved questions:*

- [*Should `rust-src` be a default component?*][unresolved-rust-src]

## Panic strategies
[panic-strategies]: #panic-strategies

> Panic strategies are unlike other profile settings insofar as they influence

Reviewer 2: Putting this here since there's not a great place to put it: while it's okay to make a future addition, I think it's good to mention the `optimize_for_size` feature as one of the other main things that people care about when building `std`. There are potential flags we could check for whether to enable this feature, but it's probably going to want its own dedicated flag associated with `build-std` and thus we shouldn't infer it here.

For example, I can imagine `-O2` plus `optimize_for_size` being useful because you're willing to permit, for example, loop unrolling or autovectorization, but not large tables being included in the binary. So, simply checking for `-Os` or `-Oz` is not adequate.

which crates are built and which flags are passed to the standard library build.
For example, if `panic = "unwind"` were set in the Cargo profile then the
`panic_unwind` feature would need to be provided to `std` and `-Cpanic=unwind`
passed to suggest that the compiler use that panic runtime.

If Cargo is not building `std`, then neither of the panic runtimes will be
built. In this circumstance rustc will continue to throw an error when a
unwinding panic strategy is chosen.

If Cargo would build `std` for a project then Cargo's behaviour depends on
whether or not `panic` is set in the profile:

- If `panic` is not set in the profile then unwinding may still be the default
  for the target and Cargo will need to enable the `panic_unwind` feature to the
  `sysroot` crate to build `panic_unwind` just in case it is used

- If `panic` is set to "unwind" then the `panic_unwind` feature of `sysroot`
  will be enabled and `-Cpanic=unwind` will be passed

- If `panic` is set to "abort" then `-Cpanic=abort` will be passed

  - `panic_abort` is a non-optional dependency of `std` so it will always be
    built

> 

Reviewer 2: Worth commenting on `-Cpanic=immediate-abort` which was recently added: rust-lang/rust#146974

`external-mem` feature. Most `no_std` users will use the `compiler_builtins`
implementation of these symbols and will work by default when they do not depend
on `std`.

Those users providing their own mem symbols can override on weak linkage of the
`compiler_builtins` symbols, or use a nightly toolchain to enable the
`external-mem` feature of an explicit dependency on the standard library (per
[*Standard library dependencies*][deps]).

*See the following sections for rationale/alternatives:*

- [*Why not rely on weak linkage for `compiler-builtins/mem` symbols?*][rationale-no-weak-linkage]
- [*Why invert the `mem` feature?*][rationale-no-mem]

## `profiler-builtins`
[profiler-builtins]: #profiler-builtins

`profiler-builtins` will not be built by build-std, thus preventing
profile-guided optimisation with a locally-built standard library.
`profiler-builtins` has native dependencies which may fail compilation of the
standard library if missing were `profiler-builtins` to be built by default as
part of the standard library build.

*See the following sections for future possibilities:*

- [*Allow building `profiler-builtins`*][future-profiler-builtins]

> ## Caching

Reviewer: `build-dir` does not change the situation regarding caching, only where they are located.  A user can force `build-dir` to be shared like they can for `target-dir` today but it is not expected to be a default or generally encouraged pattern.

standard library. [cargo#5931] tracks the feature request of intermediate
artifact caching in Cargo.

↩ [*Caching*][caching]

## Why not link to hosted standard library documentation in generated docs?
[rationale-generated-docs]: #why-not-link-to-hosted-standard-library-documentation-in-generated-docs

Cargo would need to pass `-Zcrate-attr="doc(html_root_url=..)"` to the standard
library crates when building them but doesn't have the required information to
know what url to provide. Cargo would require knowledge of the current toolchain
channel to build the correct url and doesn't know this.

↩ [*Generated documentation*][generated-documentation]

# Unresolved questions
[unresolved-questions]: #unresolved-questions

The following are aspects of the proposal which warrant further discussion or
small details are likely to be bikeshed prior to this part of the RFC's
acceptance or stabilisation and aren't pertinent to the overall design:

## What should the `build-std` configuration in `.cargo/config` be named?
[unresolved-config-name]: #what-should-the-build-std-configuration-in-cargoconfig-be-named

What should this configuration option be named? `build-std`?

> `rebuild-standard-library`?

Reviewer: I thought we had notes on considerations for this?

For instance, `build-std = "always"` could be misinterpreted as "don't apply rebuild rules"

[unresolved-config-values]: #what-should-the-always-and-never-values-of-build-std-be-named

What is the most intuitive name for the values of the `build-std` setting?
`always`? `manual`? `unconditional`?

`always` combined with the configuration option being named `build-std` -
`build-std = "always"` - is imperfect as it reads as if the standard library
will be re-built every time, when it actually just avoids use of the pre-built
standard library and caches the newly-built standard library.

↩ [*Proposal*][proposal]

## What should `build-std-crates` be named?
[unresolved-build-std-crate-name]: #what-should-build-std-crates-be-named

What should this configuration option be named?

↩ [*Proposal*][proposal]

## Should the standard library inherit RUSTFLAGS?
[unresolved-inherit-rustflags]: #should-the-standard-library-inherit-rustflags

Existing designs for *[Opaque dependencies]* intended that `RUSTFLAGS` would not
apply to the opaque dependency. However, if a target modifier were set using
`RUSTFLAGS` and build-std ignored the variable, then rustc would fail to build
the user's project due to incompatible target modifiers. This would necessitate

> that every stable target modifier be exposed via Cargo to be usable in practice.

Reviewer 3: I have run into a friction point with the current build-std in that setting `RUSTFLAGS=-Dwarnings` sometimes causes the std build to fail on T3 targets. It may be worth noting somewhere that std's lints will be suppressed and not be affected by RUSTFLAGS, similar to other dependencies.

Author: I've mentioned that we'll cap-lints for standard library crates in 4db3d98.

When `build-std` is set to "always", then the standard library will be
unconditionally recompiled ([?][rationale-unconditional]) in the release profile
defined in its workspace as part of every clean build
([?][rationale-release-profile]). This is primarily useful for users of tier
three targets. As with other dependencies, the standard library's build will
respect the `RUSTFLAGS` environment variable.

> [!NOTE]
>
> Configuration of the pre-built standard library is split across bootstrap and
> the Cargo packages for the standard library. As much of this configuration as
> possible should be moved to the Cargo profile for these packages so that the
> artifacts produced by build-std match the pre-built standard library as much
> as is feasible.

`build-std` is a short-hand for an object which sets the `when` key:

```toml
[build]
build-std = { when = "always" }
```

Alongside `build-std`, a `build-std.crates` key will be introduced
([?][rationale-build-std-crate]), which can be used to specify which crates from
the standard library should be built. Only "core", "alloc" and "std" are valid
values for `build-std.crates`.

```toml
[build]

> build-std = { when = "always", crates = "std" }

Reviewer: I would like to see the confusing name of a plural for a single value at least brought up in the "what to name" so we don't lose track of what the concern is.

> build-std = { when = "always", crates = "std" }

Reviewer 4: Another possible alternative name: `which`, to go with `when`.

`up-to` feels really quirky.

```

A value of "std" means that every crate in the graph has a direct dependency on
`std`, `alloc` and `core`. Similarly, "alloc" means `alloc` and `core`, and
"core" means just `core`.

If `std` is to be built and Cargo is building a test or benchmark using the
default test harness then Cargo will also build the `test` crate.

If [*Standard library dependencies*][rfcs#3875] are implemented then `builtin`
dependencies will be used if `build-std.crates` is not explicitly set.
Otherwise, `build-std.crates` will default to the crate intended to be supported
by the target (see later
[*Standard library crate stability*][standard-library-crate-stability] section).

> [!NOTE]
>

> > Inspired by the concept of [opaque dependencies][Opaque dependencies], the

Reviewer: @nicoburns brought up on reddit the idea of `cargo clean` not affecting opaque dependencies, requiring some other step to clean them (maybe a `cargo clean -p std`).  Alternatively, we may get this "for free" with rust-lang/cargo#5931.

I lean towards keeping `cargo clean` the same for now but might be worth calling this question out somewhere.

> or by checking if a file exists at the path matching the provided target name.
>
> This does not require any changes to rustc. If it is invoked to build the
> standard library then it will continue to do so, as is possible today, it is
> only the build-std functionality in Cargo that will not support custom targets
> initially.

Furthermore, custom targets will be destabilised in rustc (as in [rust#71009]).
This will not be a significant breaking change as custom targets cannot
effectively be used currently without nightly (needing build-std to have
`core`).

Custom targets can still be used with build-std on nightly toolchains provided
that `-Zunstable-options` is provided to Cargo.

*See the following sections for rationale/alternatives:*

- [*Why disallow custom targets?*][rationale-disallow-custom-targets]

*See the following sections for future possibilities:*

- [*Allow custom targets with build-std*][future-custom-targets]

## Preventing implicit sysroot dependencies
[preventing-implicit-sysroot-dependencies]: #preventing-implicit-sysroot-dependencies

Cargo will pass a new flag to rustc which will prevent rustc from loading

> top-level dependencies from the sysroot ([?][rationale-root-sysroot-deps]).

Reviewer 5: build-std would recompile the entire standard library, right? So there are no dependencies that need to be loaded from the sysroot.

> top-level dependencies from the sysroot ([?][rationale-root-sysroot-deps]).

Reviewer 6: https://github.com/rust-lang/rfcs/pull/3874#discussion_r2534912578

`up-to`name making sense suggests to me that it is possible to rebuild `std` without rebuilding `core`+`alloc`?

> top-level dependencies from the sysroot ([?][rationale-root-sysroot-deps]).

Reviewer 5: The way I interpret is that `up-to = "std"` would rebuild `core`, `alloc` and `std`, but not any later crates like `proc_macro` or `test`.

crates to be built for compilation to succeed.

> [!NOTE]
>
> The `unwind` crate will continue to link to the system's `libunwind` which
> will need to match the target modifiers used by the standard library to avoid
> incompatibilities. Likewise, if `llvm-libunwind`, `-Clink-self-contained=yes`
> or `-Ctarget-feature=+crt-static` are used and the distributed `libunwind` is
> used then it will also need to match the target modifiers of the standard
> library to avoid incompatibilities.

*See the following sections for future possibilities:*

- [*Avoid building `panic_unwind` unnecessarily*][future-panic_unwind]

## Building the standard library on a stable toolchain
[building-the-standard-library-on-a-stable-toolchain]: #building-the-standard-library-on-a-stable-toolchain

rustc will automatically assume `RUSTC_BOOTSTRAP` when the source path of the
crate being compiled is within the same sysroot as the rustc binary being
invoked ([?][rationale-implied-bootstrap]). Cargo will not need to use
`RUSTC_BOOTSTRAP` when compiling the standard library with a stable toolchain.
The standard library's dependencies will not be permitted to use build probes to

> detect whether a nightly version is being used.

Reviewer 5: I went through all build scripts in dependencies we have. The libc crate detects both the rustc version and if it is a nightly version. It seems like it doesn't actually use the latter for anything though. The object crate also detects rustc version. Nothing else seems to be detecting the rustc version that I could see.

Author: I've added a note that `libc` and `object` are the only affected crates at the moment in a0977f78

profile-guided optimisation with a locally-built standard library.
`profiler-builtins` has native dependencies which may fail compilation of the
standard library if missing were `profiler-builtins` to be built by default as
part of the standard library build.

*See the following sections for future possibilities:*

- [*Allow building `profiler-builtins`*][future-profiler-builtins]

### Caching
[caching]: #caching

Standard library artifacts built by build-std will be reused equivalently to
today's crates/dependencies that are built within a shared target directory. By
default, this limits sharing to a single workspace ([?][rationale-caching]).

*See the following sections for rationale/alternatives:*

- [*Why not globally cache builds of the standard library?*][rationale-caching]

### Generated documentation
[generated-documentation]: #generated-documentation

When running `cargo doc` for a project to generate documentation and rebuilding
the standard library, the generated documentation for the user's crates will
link to the locally generated documentation for the `core`, `alloc` and `std`
crates, rather than the upstream hosted generation as is typical for non-locally

> built standard libraries.

Reviewer 7: This seems like an OK compromise to start, but I'm not entirely comfortable with it. The standard library documentation takes half a gigabyte of disk space, and can take quite a while to generate.

In the future we may want to improve this situation somehow. Cargo could use the same logic that rustdoc uses to find the correct channel (I think it's around [here](https://github.com/rust-lang/rust/blob/4fa80a5e733e2202d7ca4c203c2fdfda41cfe7dc/src/librustdoc/clean/utils.rs#L580-L586)). Cargo already generates channel-specific links to doc.rust-lang.org. Or maybe there is some extension to rustdoc that we could use.

For context, the issue tracking this problem is https://github.com/rust-lang/wg-cargo-std-aware/issues/88.

Cargo also has an unstable [rustdoc-map](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#rustdoc-map) feature for controlling where the std docs link to. Perhaps we could lean in on that. Unfortunately it's stuck on a bunch of issues, and one important one is [the precedence of `--extern-html-root-url` and `--extern-html-root-takes-precedence`](https://github.com/rust-lang/rust/pull/82776).

Author: I very much intend what's written to be the conservative "this will definitely work" approach and would be happy to see us try and do something better here.

this is already impossible with the pre-built standard library.

See
[*Why vendor the standard library's dependencies?*][rationale-vendoring]

↩ [*Proposal*][proposal]

## Why not build the standard library in incremental?
[rationale-incremental]: #why-not-build-the-standard-library-in-incremental

The standard library sources are not intended to be modified locally, similarly
to those Cargo fetches from `registry` or `git` sources. Incremental compilation
would only add a compilation time overhead for any package sources which do not
change.

↩ [*Proposal*][proposal]

## Why not produce a `dylib` for the standard library?
[rationale-no-dylib]: #why-not-produce-a-dylib-for-the-standard-library

The standard library supports being built as both a `rlib` and a `dylib` and
both are shipped as part of the `rust-std` component. As the `dylib` does not
contain a metadata hash, it can be rebuilt unnecessarily when toolchain versions
change (e.g. switching between stable and nightly and back). The `dylib` is only
linked against when `-Cprefer-dynamic` is used. build-std will initially be
conservative and not include the `dylib` and `-Cprefer-dynamic` would fail

> compilation.

Reviewer 5: I'm pretty sure `-Cprefer-dynamic` would fallback to static linking rather than failing compilation.

difference in how a crate is made available to rustc should not be observable to
the user as they have not opted into the migration.

Passing crates without `noprelude` with the existing prelude behaviour has also
been a source of [bugs][wg-cargo-std-aware#40] in previous `-Zbuild-std`
implementations.

↩ [*Preventing implicit sysroot dependencies*][preventing-implicit-sysroot-dependencies]

[wg-cargo-std-aware#40]: https://github.com/rust-lang/wg-cargo-std-aware/issues/40

## Why not allow the source path for the standard library be customised?
[rationale-custom-src-path]: #why-not-allow-the-source-path-for-the-standard-library-be-customised

It is not a goal of this proposal to enable or improve the usability of custom
or modified standard libraries.

↩ [*Vendored `rust-src`*][vendored-rust-src]

## Why vendor the standard library's dependencies?
[rationale-vendoring]: #why-vendor-the-standard-librarys-dependencies

Vendoring the standard library is possible since it currently has its own
workspace, allowing the dependencies of just the standard library crates (and
not the compiler or associated tools in `rust-lang/rust`) to be easily packaged.

> Doing so has multiple advantages..

Reviewer 7: Yea, I'm not sure how we're going to implement this. I'm a little concerned it will be tricky, and is good for us to keep in mind that this will be a challenging change.

stable toolchains, it is worth establishing a narrow general mechanism.

For example, Rust's project goal to enable Rust for Linux to build using only a
stable toolchain would require that it be possible to build `core` without
nightly.

It is not sufficient for rustc to special-case the `core`, `alloc` and `std`
crate names as, when being built as part of the standard library, dependencies
of the standard library also use unstable features and it is not practical to
special-case all of these crates.

↩ [*Building the standard library on a stable toolchain*][building-the-standard-library-on-a-stable-toolchain]

## Why invert the `mem` feature?
[rationale-no-mem]: #why-invert-the-mem-feature

While "negative" features are typically discouraged due to how features unify
(e.g. `std` features are preferred to `no_std`): the `mem` feature's current
behaviour is the opposite of what is optimal.

Ideally, a crate should be able to provide alternate memory symbols and disable
`compiler_builtins`' symbols for the entire crate graph by enabling a feature
(e.g. `std`/`libc` could do this) - this is what an `external-mem` feature

> enables.

Reviewer 5: Until external-mem automatically gets enabled for all targets with a libc implementation, this inversion of the mem feature would cause a silent performance hit for those targets, while previously you would get a loud failure, which IMO is preferable.

([?][rationale-build-std-precedence]):

1. `[target.<triple>]`
2. `[target.<cfg>]`
3. `[build]`

As the Cargo configuration is local to the current user (typically in
`.config/cargo.toml` in the project root and/or Cargo home directory), the value
of `build-std` is not influenced by the dependencies of the current crate.

When `build-std` is set to "always", then the standard library will be
unconditionally recompiled ([?][rationale-unconditional]) in the release profile
defined in its workspace as part of every clean build
([?][rationale-release-profile]). This is primarily useful for users of tier
three targets. As with other dependencies, the standard library's build will
respect the `RUSTFLAGS` environment variable.

> [!NOTE]
>
> Configuration of the pre-built standard library is split across bootstrap and
> the Cargo packages for the standard library. As much of this configuration as
> possible should be moved to the Cargo profile for these packages so that the
> artifacts produced by build-std match the pre-built standard library as much
> as is feasible.

`build-std` is a short-hand for an object which sets the `when` key:

```toml
[build]
build-std = { when = "always" }

> ```

Reviewer: Something that came up with `config-include` is deciding whether the primary form of a table should be a standard table or inline table and that can help us decide whether short-hands are appropriate or not.

Is there a reason you are using an inline table rather than a standard table?

Author: Not especially, just for conciseness in the RFC text. My gut feeling is that `build-std.when` would be most common because it'll eventually be the only option that users need to set.

Alongside `build-std`, a `build-std.crates` key will be introduced
([?][rationale-build-std-crate]), which can be used to specify which crates from
the standard library should be built. Only "core", "alloc" and "std" are valid
values for `build-std.crates`.

```toml
[build]
build-std = { when = "always", crates = "std" }
```

A value of "std" means that every crate in the graph has a direct dependency on
`std`, `alloc` and `core`. Similarly, "alloc" means `alloc` and `core`, and
"core" means just `core`.

If `std` is to be built and Cargo is building a test or benchmark using the
default test harness then Cargo will also build the `test` crate.

If [*Standard library dependencies*][rfcs#3875] are implemented then `builtin`
dependencies will be used if `build-std.crates` is not explicitly set.
Otherwise, `build-std.crates` will default to the crate intended to be supported
by the target (see later

> [*Standard library crate stability*][standard-library-crate-stability] section).

Reviewer: This reads in a confusing way where the higher precedent item is given in an "otherwise it will default" branch of the logic.

> the `sysroot` crate of the standard library workspace. `alloc` and `std` will
> be optional dependencies of the `sysroot` crate which will be enabled when the
> user has requested them. Panic runtimes are dependencies of `std` and will be
> enabled depending on the features that Cargo passes to `std` (see
> [*Panic strategies*][panic-strategies]).
>
> rustc loads panic runtimes in a different way to most dependencies, and
> without looking in the sysroot they will fail to load correctly unless passed
> in with `--extern`. rustc will need to be patched to be able to load panic
> runtimes from `-L dependency=` paths in line with other transitive
> dependencies.
>
> The standard library will always be a non-incremental build
> ([?][rationale-incremental]), Cargo's dep-info fingerprint tracking will not
> track the standard library crate sources, Cargo's `.d` dep-info file will not
> include standard library crate sources, and only a `rlib` produced (no
> `dylib`) ([?][rationale-no-dylib]). It will be built in the Cargo `target`
> directory of the crate or workspace like any other dependency.
>
> Standard library crates are provided to the compiler using the `--extern` flag

> > with the `noprelude` modifier ([?][rationale-noprelude-with-extern]).

Author: Mentioned this in 39738034

Always using a locally-built standard library avoids the complexity associated
with an automatic build-std mechanism while still being useful for users of tier
three targets. By leaving an automatic mechanism for a later RFC, fewer of the
technical challenges of build-std need to be addressed all at once.

Having an opt-in mechanism initially, such as `build-std.when = "always"`,
allows for early issues with build-std to be ironed out without potentially
affecting more users like an automatic mechanism. Later proposals will extend
the `build-std` option with an automatic mechanism.

↩ [*Proposal*][proposal]

### Why does "always" rebuild in release profile?
[rationale-release-profile]: #why-does-always-rebuild-in-release-profile

The release profile most closely matches the existing pre-built standard
library, which has proven itself suitable for a majority of use cases.

By minimising the differences between a newly-built std and a pre-built std,
there is less chance of the user experiencing bugs or unexpected behaviour from
the well-tested and supported pre-built std. Later proposals will extend the
`build-std` option with customised standard library builds that use the user's

> profile.

Reviewer: I wish I had written more about this, how much is build-std specific vs opaque dependencies.  In general, I want us to be making this just a built-in opaque dependency and there to be principled ways of applying these ideas.

`always`? `manual`? `unconditional`?

`always` combined with the configuration option being named `build-std` -
`build-std.when = "always"` - is imperfect as it reads as if the standard
library will be re-built every time, when it actually just avoids use of the
pre-built standard library and caches the newly-built standard library.

↩ [*Proposal*][proposal]

### What should `build-std.crates` be named?
[unresolved-build-std-crate-name]: #what-should-build-stdcrates-be-named

What should this configuration option be named? In particular, `crates` being
plural is unintuitive: while it can enable building multiple crates, it is set
with the name of only a single crate. `up-to` might be better.

↩ [*Proposal*][proposal]

### Should the standard library inherit RUSTFLAGS?
[unresolved-inherit-rustflags]: #should-the-standard-library-inherit-rustflags

Existing designs for *[Opaque dependencies]* intended that `RUSTFLAGS` would not
apply to the opaque dependency. However, if a target modifier were set using
`RUSTFLAGS` and build-std ignored the variable, then rustc would fail to build
the user's project due to incompatible target modifiers. This would necessitate

> that every stable target modifier be exposed via Cargo to be usable in practice.

Reviewer: When talking about target modifiers, it is an exception and one I would not be thrilled with stabilizing.  If we're taking the stance though that we're blindly applying RUSTFLAGS to opaque dependencies and the way to avoid that is to find ways to model RUSTFLAGS in other parts of Cargo, like profiles, then that is hopefully good enough and not something we'll come to regret.

> standard library is resolved differently to other dependencies:
>
> - The lockfile included in the standard library source will be used when
>   resolving the standard library's dependencies ([?][rationale-lockfile]).
>
> - The dependencies of the standard library crates are entirely opaque to the
>   user. Different semver-compatible versions of these dependencies can
>   exist in the user's resolve. The user cannot control compilation any of
>   the dependencies of the `core`, `alloc` or `std` standard library crates
>   individually (via profile overrides, for example).
>
> - The release profile defined by the standard library will be used.
>
>     - This profile will be updated to match the current compilation options
>       used by the pre-built standard library as much as possible (e.g. using
>       `-Cembed-bitcode=yes` to support LTO).
>
> - Standard library crates and their dependencies from `build-std.crates`
>   cannot be patched/replaced by the user in the Cargo manifest or config
>   (e.g. using source replacement, `[replace]` or `[patch]`)
>
> - Lints in standard library crates will be built using `--cap-lints allow`
>   matching other upstream dependencies.
>

> > Cargo will resolves the dependencies of opaque dependencies, such as the

Reviewer 4: ```suggestion
> Cargo will resolve the dependencies of opaque dependencies, such as the
```

Author: Fixed in 801e1e35

> user has requested them. Panic runtimes are dependencies of `std` and will be
> enabled depending on the features that Cargo passes to `std` (see
> [*Panic strategies*][panic-strategies]).
>
> rustc loads panic runtimes in a different way to most dependencies, and
> without looking in the sysroot they will fail to load correctly unless passed
> in with `--extern`. rustc will need to be patched to be able to load panic
> runtimes from `-L dependency=` paths in line with other transitive
> dependencies.
>
> The standard library will always be a non-incremental build
> ([?][rationale-incremental]), Cargo's dep-info fingerprint tracking will not
> track the standard library crate sources, Cargo's `.d` dep-info file will not
> include standard library crate sources, and only a `rlib` produced (no
> `dylib`) ([?][rationale-no-dylib]). It will be built in the Cargo `target`
> directory of the crate or workspace like any other dependency.
>
> Standard library crates are provided to the compiler using the `--extern` flag
> with the `noprelude` modifier ([?][rationale-noprelude-with-extern]). Standard
> library crates are also provided with the `nounused` modifier to avoid being

> > considered an unused crate dependency.

Reviewer 4: Please consider adding a rationale for `nounused`.

Author: Added in 9171316d

*See the following sections for relevant unresolved questions:*

- [*What should the `build-std` configuration in `.cargo/config` be named?*][unresolved-config-name]
- [*What should the "always" and "never" values of `build-std` be named?*][unresolved-config-values]
- [*What should `build-std.crates` be named?*][unresolved-build-std-crate-name]
- [*Should the standard library inherit RUSTFLAGS?*][unresolved-inherit-rustflags]

*See the following sections for future possibilities:*

- [*Adding a shorthand*][future-shorthand]
- [*Allow reusing sysroot artifacts if available*][future-reuse-sysroot]

[Opaque dependencies]: https://github.com/rust-lang/cargo/issues/3573#issuecomment-3498262549

### Standard library crate stability
[standard-library-crate-stability]: #standard-library-crate-stability

An optional `standard_library_support` field
([?][rationale-why-standard-library-support]) is added to the target
specification ([?][rationale-target-spec-purpose]), replacing the existing
`metadata.std` field. `standard_library_support` has two fields:

- `supported`, which can be set to either "core", "core and alloc", or

>   "core, alloc, and std"

Reviewer 4: ```suggestion
- `supported`, which can be set to either "core", "core, alloc", or
  "core, alloc, std"
```

Author: Fixed in 801e1e35

- `default`, which can be set to either "core", "core and alloc", or
  "core, alloc, and std"
  - `default` cannot be set to a value which is "less than" that of `supported`

>     (i.e. "core and alloc" when `supported` was only set to "core")

Reviewer 4: ```suggestion
- `default`, which can be set to either "core", "core,  alloc", or
  "core, alloc, std"
  - `default` cannot be set to a value which is "less than" that of `supported`
    (i.e. "core, alloc" when `supported` was only set to "core")
```

Author: Fixed in 801e1e35

The `supported` field determines which standard library crates Cargo will permit
to be built for this target on a stable toolchain. On a nightly toolchain, Cargo
will build whichever standard library crates are requested by the user.

The `default` field determines which crate will be built by Cargo if
`build-std.when = "always"` and `build-std.crates` is not set. Users can specify
`build-std.crates` to build more crates than included in the `default`, as long
as those crates are included in `supported`.

The correct value for `standard_library_support` is independent of the tier of
the target and depends on the set of crates that are intended to work for a
given target, according to its maintainers.

If `standard_library_support` is unset for a target, then Cargo will not permit
any standard library crates to be built for the target on a stable toolchain. It
will be required to use a nightly toolchain to use build-std with that target.

Cargo's `build-std.crates` field will default to the value of the
`standard_library_support.default` field (`std` for "core, alloc, and std",

> `alloc` for "core and alloc", and `core` for "core"). This does not prevent

Reviewer 4: ```suggestion
`standard_library_support.default` field (`std` for "core, alloc, std",
`alloc` for "core, alloc", and `core` for "core"). This does not prevent
```

Author: Fixed in 801e1e35

artifact caching in Cargo.

↩ [*Caching*][caching]

[cargo#5931]: https://github.com/rust-lang/cargo/issues/5931

### Why not link to hosted standard library documentation in generated docs?
[rationale-generated-docs]: #why-not-link-to-hosted-standard-library-documentation-in-generated-docs

Cargo would need to pass `-Zcrate-attr="doc(html_root_url=..)"` to the standard
library crates when building them but doesn't have the required information to
know what url to provide. Cargo would require knowledge of the current toolchain
channel to build the correct url and doesn't know this.

↩ [*Generated documentation*][generated-documentation]

## Unresolved questions
[unresolved-questions]: #unresolved-questions

The following are aspects of the proposal which warrant further discussion or
small details are likely to be bikeshed prior to this part of the RFC's
acceptance or stabilisation and aren't pertinent to the overall design:

### What should the `build-std.when` configuration in `.cargo/config` be named?
[unresolved-config-name]: #what-should-the-build-stdwhen-configuration-in-cargoconfig-be-named

What should this configuration option be named? `build-std`?

> `rebuild-standard-library`?

Reviewer 4: I think `when` is a fine name.

↩ [*Proposal*][proposal]

### What should the "always" and "never" values of `build-std` be named?
[unresolved-config-values]: #what-should-the-always-and-never-values-of-build-std-be-named

What is the most intuitive name for the values of the `build-std` setting?
`always`? `manual`? `unconditional`?

`always` combined with the configuration option being named `build-std` -
`build-std.when = "always"` - is imperfect as it reads as if the standard
library will be re-built every time, when it actually just avoids use of the
pre-built standard library and caches the newly-built standard library.

↩ [*Proposal*][proposal]

### What should `build-std.crates` be named?
[unresolved-build-std-crate-name]: #what-should-build-stdcrates-be-named

What should this configuration option be named? In particular, `crates` being
plural is unintuitive: while it can enable building multiple crates, it is set

> with the name of only a single crate. `up-to` might be better.

Reviewer 4: ```suggestion
with the name of only a single crate. `up-to` or `which` might be better.
```

Author: Added in 801e1e35

section ([?][rationale-build-std-in-config]), with a `when` key permitting one
of two values - "never" ([?][rationale-build-std-never]) or "always", defaulting
to "never":

```toml
[build]
build-std.when = "never" # or `always`
```

`build-std` can also be specified in the `[target.<triple>]` and
`[target.<cfg>]` sections ([?][rationale-build-std-target-section]):

```toml
[target.aarch64-unknown-illumos]
build-std.when = "never" # or `always`
```

The `build-std` configuration locations have the following precedence
([?][rationale-build-std-precedence]):

1. `[target.<triple>]`
2. `[target.<cfg>]`
3. `[build]`

As the Cargo configuration is local to the current user (typically in
`.config/cargo.toml` in the project root and/or Cargo home directory), the value
of `build-std` is not influenced by the dependencies of the current crate.

When `build-std` is set to "always", then the standard library will be
unconditionally recompiled ([?][rationale-unconditional]) in the release profile
([?][rationale-release-profile]) defined in its workspace as part of every clean
build. This is primarily useful for users of tier three targets. As with other

> dependencies, the standard library's build will respect the `RUSTFLAGS`

Reviewer 8: When we say `RUSTFLAGS`, I guess we refer to [all different stable ways to specify rustflags in Cargo](https://doc.rust-lang.org/nightly/cargo/reference/config.html#buildrustflags), their precedence, right?

> * env `CARGO_ENCODED_RUSTFLAGS`
> * env `RUSTFLAGS`
> * config `target.<triple>.rustflags` (including its env equivalent)
> * config `target.<cfg>.rustflags` (including its env equivalent)
> * config `build.rustflags (including its env equivalent)

Reviewer 8: (As we don't respect profile right now, `profile.rustflags` unstable features won't be applied. Also we can leave interactions with other nightly features `target-applies-to-host` for future)

Author: I think it makes sense that this would apply to all of the ways that `RUSTFLAGS` gets set, and that `profile.rustflags` wouldn't apply (from the user's profile, but it probably should from the profile that the build uses - i.e. the standard library's release profile). Fixed in b87180cb.

Reviewer 8: Thanks for the update!

- [*Why accept `never` as a value for `build-std`?*][rationale-build-std-never]
- [*Why add `build-std` to the `[target.<triple>]` and `[target.<cfg>]` sections?*][rationale-build-std-target-section]
- [*Why does `[target]` take precedence over `[build]` for `build-std`?*][rationale-build-std-precedence]
- [*Why does "always" rebuild unconditionally?*][rationale-unconditional]
- [*Why does "always" rebuild in release profile?*][rationale-release-profile]
- [*Why add `build-std.crates`?*][rationale-build-std-crate]
- [*Why use the lockfile of the `rust-src` component?*][rationale-lockfile]
- [*Why not build the standard library in incremental?*][rationale-incremental]
- [*Why not produce a `dylib` for the standard library?*][rationale-no-dylib]
- [*Why use `noprelude` with `--extern`?*][rationale-noprelude-with-extern]
- [*Why use the pre-built standard library for procedural macros and build scripts in host mode?*][rationale-host-deps-host]
- [*Why use the pre-built standard library for procedural macros and build scripts in cross-compile mode?*][rationale-host-deps-cross]

*See the following sections for relevant unresolved questions:*

> - [*What should the `build-std` configuration in `.cargo/config` be named?*][unresolved-config-name]

Reviewer 8: nit: `.cargo/config` (without `.toml` extension`) is not preferred

```suggestion
- [*What should the `build-std` configuration in `.cargo/config.toml` be named?*][unresolved-config-name]
```

Author: Fixed in 237f16e3

on `std`.

Those users providing their own mem symbols can override on weak linkage of the
`compiler_builtins` symbols, or use a nightly toolchain to enable the
`external-mem` feature of an explicit dependency on the standard library (per
[*Standard library dependencies*][rfcs#3875]).

*See the following sections for rationale/alternatives:*

- [*Why not rely on weak linkage for `compiler-builtins/mem` symbols?*][rationale-no-weak-linkage]
- [*Why invert the `mem` feature?*][rationale-no-mem]

### `profiler-builtins`
[profiler-builtins]: #profiler-builtins

`profiler-builtins` will not be built by build-std, thus preventing
profile-guided optimisation with a locally-built standard library.
`profiler-builtins` has native dependencies which may fail compilation of the
standard library if missing were `profiler-builtins` to be built by default as
part of the standard library build.

*See the following sections for future possibilities:*

- [*Allow building `profiler-builtins`*][future-profiler-builtins]

### Caching
[caching]: #caching

Standard library artifacts built by build-std will be reused equivalently to

> today's crates/dependencies that are built within a shared target directory. By

Reviewer 8: This might be an implementation detail open to change after this RFC merge, though I wonder whether `libstd.rlib` is a final artifact or intermediate artifacts. As noted earlier in some review comment, Cargo now splits out build-directory (intermediate artifacts, not accessible by end user and its layout is unstable) from target directory.

I guess we consider it intermediate artifact, right? So this RFC will not allow user to abuse `build-std` as a way to build and distribute libstd artifacts.

Reviewer 5: If you dynamically link libstd it did have to be uplifted, otherwise I agree it should be intermediate artifacts.

Author: Agreed with both - noted in c55d12f.
